### ISSUE TITLE
Any idea what is causing the final pipeline to have f1 of 0?

### ISSUE THREAD

**Issue by ben-ix** [laic-ufmg/Recipe#5]
Hi there,

I'm running a comparison against a range of datasets. Most (14 others) worked fine but I am having troubles with [primary tumor ](https://archive.ics.uci.edu/ml/datasets/primary+tumor) and [post operative](https://archive.ics.uci.edu/ml/datasets/Post-Operative+Patient).  I have also attached sample files (renamed as .txt so GitHub will accept them)

Sample files:
[Postop test](https://github.com/laic-ufmg/Recipe/files/2452867/postop-test.txt)
[Postop train](https://github.com/laic-ufmg/Recipe/files/2452868/postop-train.txt)
[Tumor test](https://github.com/laic-ufmg/Recipe/files/2452869/tumor-test.txt)
[Tumor train](https://github.com/laic-ufmg/Recipe/files/2452870/tumor-train.txt)

These ran for 8+ hours, and the best resulting pipelines look like (for all of the runs):

`Best Pipeline: SA*  GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # Ensemble  `

With a f1 of 0.0 on the test set

Any idea what could be causing this? Also if theres a more appropriate way of contacting you please let me know, I don't mean to spam the issues I am just trying to get a set of results for an upcoming conference and want to ensure they are fair comparisons

Thanks!

**Comment by walterjgsp** [comment]
Hi Ben,

There are two reasons that I can think right now just looking at your pipeline (I will still look at your datasets): one is low memory, if you have executed the experiments of the big datasets together with the small datasets, it is possible that in some generation the memory all the memory of your computer is used and the evaluation for the following generations get compromised. The other reason that I can think about is the characteristics of your datasets, you have categorical or missing values in this particular datasets?

**Comment by ben-ix** [comment]
Hi Walter,

Thank you. The data sets are categorical and may have missing values, the missing values im dealing with as [here](https://github.com/laic-ufmg/Recipe/issues/4). For the categorical values, should I perform a one hot encoding prior to running instead? The code looked to do a LabelEncoder so I left it alone

Thanks

**Comment by ben-ix** [comment]
Just ran a minimal example (on post operative) for 1 minute as I did not save the outputs from the previous runs: 
```
Processing 0 of 10000 |--------------------| 0.0% Best found: 0.0 -> SA*  GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # Gaussian
Processing 1 of 10000 |--------------------| 0.0% Best found: 0.0 -> SA*  GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # Gaussian
Processing 2 of 10000 |--------------------| 0.0% Best found: 0.0 -> SA*  GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # Gaussian
Processing 3 of 10000 |--------------------| 0.0% Best found: 0.0 -> SA*  GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # Gaussian
Processing 4 of 10000 |--------------------| 0.0% Best found: 0.0 -> SA*  GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # Gaussian
Processing 5 of 10000 |--------------------| 0.1% Best found: 0.0 -> SA*  GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # Gaussian
Processing 6 of 10000 |--------------------| 0.1% Best found: 0.0 -> SA*  GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # Gaussian
Processing 7 of 10000 |--------------------| 0.1% Best found: 0.0 -> SA*  GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # Gaussian
Processing 8 of 10000 |--------------------| 0.1% Best found: 0.0 -> SA*  GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # Gaussian
Processing 9 of 10000 |--------------------| 0.1% Best found: 0.0 -> SA*  GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # Gaussian
Processing 10 of 10000 |--------------------| 0.1% Best found: 0.0 -> SA*  GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # Gaussia
Processing 11 of 10000 |--------------------| 0.1% Best found: 0.0 -> SA*  GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # Gaussia
Processing 12 of 10000 |--------------------| 0.1% Best found: 0.0 -> SA*  GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # Gaussia
Processing 13 of 10000 |--------------------| 0.1% Best found: 0.0 -> SA*  GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # GaussianNB # Ensemble 
Stop criterion - Timeout GGP
```

The [fitmap](https://github.com/laic-ufmg/Recipe/files/2454614/fitmap.txt) appears to have all 0 fitness as well. I believe the memory is fine, particular since I just tried on a very small scale (30 pop, for 1 minute total, <100 training instances).

I'm wondering if the issue is to do with the use of the LabelEncoder.

The way this is used (in pipeline.py and testAlgorithm.py) is to fit_transform a label encoder to both the training and testing data frames separately. One potential problem is if there is data in the training set not seen in the test set, the encodings will be different (i.e. a 3 in the train will be different to a 3 in the test). Instead I think a label encoder should be fit to the training data only, then this used to transform both the train and test data.  In fact I think this could be whats happening, as a particular class may not appear in a test set, and the mappings would then be off

**Comment by alexgcsa** [comment]
Hi Ben,

Firstly, can you send us the parameters you used to run the experiments? This is just to replicate the experiments.

Looking the files you sent... Have you tried to remove the char "#"  in the beginning of the file? Do the other datasets have this "#" ? Perhaps, this is causing a problem.

Cheers,

Alex de Sá

**Comment by alexgcsa** [comment]
Answering this part:

"The way this is used (in pipeline.py and testAlgorithm.py) is to fit_transform a label encoder to both the training and testing data frames separately. One potential problem is if there is data in the training set not seen in the test set, the encodings will be different (i.e. a 3 in the train will be different to a 3 in the test). Instead I think a label encoder should be fit to the training data only, then this used to transform both the train and test data. In fact I think this could be whats happening, as a particular class may not appear in a test set, and the mappings would then be off. What are your thoughts?"


You are right. This can happen. I would suggest to you to perform stratified sampling on your data to split it into train and test. This will ensure similar class distributions on both sets.

**Comment by alexgcsa** [comment]
Ben,

I believe I have just found the problem here. The problem is that your dataset must have at least two examples for each class label in the training set. This is a requirement from scikit-learn. I received this message when running the pipelines on your training set:

"The least populated class in y has only 1 member, which is too few. The minimum number of labels for any class cannot be less than 2".

For instance, in the dataset tumor-train, the class labels "class6", "class16" and "class21" have just one example representing them.
